models/ed/R/SAS.ED2.R

```diff
@@ -42,23 +19,180 @@
+  # Soil moisture potential at saturation [ m ]
+  -1. * (10.^(2.17 - 0.63*slxclay - 1.58*slxsand)) * 0.01
+  
+}
+calc.slbs   <- function(slxsand, slxclay){
+  # B exponent (unitless)
+  3.10 + 15.7*slxclay - 0.3*slxsand
+}
+
+calc.soilcp <- function(slmsts, slpots, slbs){         
+  # Dry soil capacity (at -3.1MPa) [ m^3/m^3 ].
+  # soil(nslcon)%soilcp  = soil(nslcon)%slmsts                                        &
+  #   *  ( soil(nslcon)%slpots / (soilcp_MPa * wdns / grav))       &
+  #   ** (1. / soil(nslcon)%slbs)
+  soilcp_MPa = -3.1
+  wdns = 1.000e3    
+  grav=9.80665
+  
+  slmsts *  (slpots / (soilcp_MPa * wdns / grav)) ^ (1./slbs)
+}
+
+# How Ed calculates the fire threshold if sm_fire < 0
+smfire.neg <- function(slmsts, slpots, smfire, slbs){
+  grav=9.80665
+  # soil(nsoil)%soilfr = soil(nsoil)%slmsts * Â ( soil(nsoil)%slpots / (sm_fire * 1000. / grav)) ** ( 1. / soil(nsoil)%slbs)
+  # soilfr = slmsts * ( slpots / (sm_fire * 1000. / grav)) ** ( 1. / slbs)
+  soilfr <- slmsts*((slpots/(smfire * 1000/9.80665))^(1/slbs))
+  
+  return(soilfr)
+}
+
+# How Ed calculates the fire threshold if sm_fire > 0
+smfire.pos <- function(slmsts, soilcp, smfire){
+  soilfr <- soilcp + smfire * (slmsts - soilcp)
+  return(soilfr)
+}
+
+
+##' @name SAS.ED2
+##' @title Use semi-analytical solution to accellerate model spinup
+##' @author Christine Rollinson, modified from original by Jaclyn Hatala-Matthes (2/18/14)
+##'         2014 Feb: Original ED SAS solution Script at PalEON modeling HIPS sites (Matthes)
+##'         2015 Aug: Modifications for greater site flexibility & updated ED
+##'         2016 Jan: Adaptation for regional-scale runs (single-cells run independently, but executed in batches)
+##'         2018 Jul: Conversion to function, Christine Rollinson July 2018
+##'@description This functions approximates landscape equilibrium steady state for vegetation and 
+##'             soil pools using the successional trajectory of a single patch modeled with disturbance
+##'             off and the prescribed disturbance rates for runs (Xia et al. 2012 GMD 5:1259-1271). 
+##' @param dir.analy Location of ED2 analyis files; expects monthly and yearly output
+##' @param dir.histo Location of ED2 history files (for vars not in analy); expects monthly
+##' @param outdir Location to write SAS .css & .pss files
+##' @param lat site latitude; used for file naming
+##' @param lon site longitude; used for file naming
+##' @param block Number of years between patch ages
+##' @param treefall Value to be used for TREEFALL_DISTURBANCE_RATE in ED2IN for full runs (disturbance on)
+##' @param sufx ED2 out file suffix; used in constructing file names(default "g01.h5) 
+##' @export
+##'
+SAS.ED2 <- function(dir.analy, dir.histo, outdir, lat, lon, block,
+                    prefix, treefall, param.args = SAS.ED2.param.Args(), sufx =
+                      "g01.h5") {
+  
+  #--- Reading in the parameters
+  decomp_scheme <- param.args[1]
```

Comment:
This will break badly if anyone passes a non-default param.args. At a minimum this should match by name instead of index, but probably better would be to use all params directly from the vector, e.g. `if(!param.args$decomp_scheme %in% 0:4) stop("Invalid decomp_scheme")` and so on.